### ISSUE TITLE
Idea: account for temp-induced noise in existing training data

### ISSUE THREAD

**Issue by kk2796** [LeelaChessZero/lc0#955]
**Rationale**

Consider a position in the training data where it is white's move, game result is Z=-1 ("black won"), and the best move after an 800-node search had 200 visits.  However, due to temp, rather than playing the 200-visit move, white instead played a (presumably much weaker) move that only got 10 visits.  While this does not utterly invalidate that game result of Z=-1, it should reduce the confidence we have that the position is truly losing for white.  Intuitively, it seems that that if the value-head scored this position as 0.0 (dead draw), that value estimate is somehow "less wrong" than if it had guessed 0.0 and the best 200-visit move were played. 

Conversely, consider the same scenario (where white is to move and - thanks to temp - chooses the same crummy 10-visit move instead of 200-visit move), except now pretend game result was Z=+1 ("white won").  This would cast the temp-induced 10-visit move into a different light.  We should now appreciate that "white won" (Z=+1) _despite_ playing the (presumably-weak) 10-visit move.   Here, the selection of such a weak move should actually increase our confidence in Z=+1.  Intuitively, "This position is so winning for white that white was able win despite being forced to play a really bad move".  If the value-head scored this position as 0.0 (dead draw), that should be considered "more wrong" in light of the 10-node move.

I suspect that Leela (and A0) have always been plagued by the inability to account for the haziness of Z values in certain positions, and propose something be done to tackle the shortcoming head-on.

**Proposal**
My idea is to incorporate this intuition into the training process with 
1) a "haziness" calculation that computes the influence of temperature on game result for all positions of existing training games.  
2) a modification to training logic which uses each position's haziness to reduce/dampen the loss function's sensitivity to value-head error for that position.  

The value-head component of loss function is cited in A0 paper as (Z-v)^2. I propose this term be multiplied by a "dampening coefficient" defined using three terms, H, HDF, and HC.
Old: (Z-v)^2
New: (Z-v)^2 * HC * ((1-HDF)+HDF*(1-H)) 

**Haziness (H)** is the key component.  The H value of a position would represent the estimated cumulative impact of temperature of moves played after that position on the Z-value of the position.  The higher the value of a position's H, the greater the estimated impact of temp on the position's Z-value. The H value of the final position of each game is defined to be 0.0 (which makes sense: there are no more moves to be played, so temp cannot have affected the Z value of the final position).  The H value of all other positions are defined inductively; the H-value of each position is based on H value of the subsequent position of that game, notched upward or downward based on the move selection leading to that subsequent position (see below for H computation algorithm).

**Haziness-Dampening-Factor(HDF)** One of the biggest risks of this strategy is that Leela might, on average, accumulate so much haziness that the majority of opening positions for most games winds up being dampened out of existence.  HDF is a backstop to the dampening (a "volume knob" for the entire idea).  Basically, HDF is the maximum amount of dampening a given position can experience, regardless of its haziness. HDF=0 disables dampening altogether; HDF=0.5 becomes the average of dampened/undampened value; HDF=1.0 means that positions with haziness H very close to 1.0 would effectively have zero impact on value-head training.

**Haziness Correction (HC)** Even with the HDF limiting the effect of dampening, it's important to note that the dampening coefficient is guaranteed to reduce the overall contribution of value-head errors to the overall loss equation.  This alters the balance between influence of policy-head on training vs influence of value-head on training.  HC simply exists to correct for this and restore that balance - the idea is to empirically measure the average level of dampening observed for the training HDF and temp and set HC to 1/avg(dampening). 

**Computing H:** 

H values of a given game are computed by first setting H=0.0 for the final position of the game (or, in the case of TB rescoring, the latest position identified as exact W/L/D).  The H values of all other positions are computed by stepping backward from that final position to starting position.  For the P'th position, denoted pos[P], the value of pos[P].H is computed as follows:

let max_n = maximum number of visits over all actions of pos[P] (i.e. the number of visits for the best move of the position)  
let sel_n = number of visits for the selected action (i.e. the move that was played at position P during training that led to susequent position, P+1)
let adjust_factor = (max_n-sel_n)/max_n
if Z of position = "I won" 
then
    set pos[P].H = pos[P+1].H + (1-pos[P+1].H)*adjust_factor
else (if Z = "I lost" or Z = "Draw")
    set pos[P].H =  pos[P+1].H - (1-pos[P+1].H)*adjust_factor

I'll use a hypothetical 7-position training game (ending with "white win" at 7th half-move) to illustrate how H propagates through the game:

1) max_n=100, sel_n=100 W to move, Z="I won"
2) max_n=80, sel_n=5 B to move Z="I lost"
3) max_n=120, sel_n=30 W to move Z="I won"
4) max_n=90, sel_n=80 B to move Z="I lost"
5) max_n=100, sel_n=40 W to move Z="I won"
6) max_n=80, sel_n=70 B to move Z="I lost"
7) W to move Z="I won"

Algorithm starts at P=7:

P=7) W to move Z="I won"
pos[7].haziness = 0.0 (by definition).

P=6) max_n=80, sel_n=70 B to move Z="I lost"
adjust_factor= (80-70)/80 = 0.125. 
Since Z="I lost"
pos[6].haziness = pos[7].haziness + (1-pos[7].haziness) * adjust_factor
pos[6].haziness = 0.0 + (1.0 - 0.0)*0.125
pos[6].haziness = 0.125
 
P=5) max_n=100, sel_n=40 W to move Z="I won"
adjust_factor = (100-40)/100 = 0.6
Since Z="I won"
pos[5].haziness = pos[6].haziness * (1-adjust_factor)
pos[5].haziness = 0.125 * (1-0.6)
pos[5].haziness = 0.05

P=4)  max_n=90, sel_n=80 B to move Z="I lost"
adjust_factor= (90-80)/90 =0.111. 
Since Z="I lost"
pos[4].haziness = pos[5].haziness + (1-pos[5].haziness) * adjust_factor
pos[4].haziness = 0.05 + (1 - 0.05)*0.111
pos[4].haziness = 0.156
 
P=3)  max_n=120, sel_n=30 W to move Z="I won"
adjust_factor = (120-30)/120=.75 
Since  Z="I won"
pos[3].haziness = pos[4].haziness * (1-adjust_factor)
pos[3].haziness =  .156 * (1-0.75)
pos[3].haziness = 0.039

P=2) max_n=80, sel_n=5 B to move Z="I lost"
adjust_factor= (80-5)/80 =0.938. 
Since Z="I lost"
pos[2].haziness = pos[3].haziness + (1-pos[3].haziness) * adjust_factor
pos[2].haziness = 0.039 + (1 - 0.039)*0.938 
pos[2].haziness = 0.94
 
P=1) max_n=100, sel_n=100W to move, Z="I won"
adjust_factor = (100-100)/100=0 
Since  Z="I won"
pos[].haziness = pos[2].haziness * (1-adjust_factor)
pos[].haziness = 0.94 * (1-0)
pos[].haziness = 0.94

Here are the dampening coefficients for each position, if we use an HDF=0.5 and HC = 1.2
P1) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0.94)) = 0.64
P2) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0.94)) = 0.64
P3) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0.039)) = 1.18
P4) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0.156)) = 1.11
P5) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0.05)) = 1.17
P6) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0.125)) = 1.13
P7) ((1-HDF)+HDF*(1-H)) = 1.2 * (0.5 + 0.5 * (1-0)) = 1.20

**Comment by Naphthalin** [comment]
I think the suggestions in this issue are outdated, as we now use cross entropy training of a WDL head. Also, this sounds very complicated without a clear advantage -- did you ever try to implement this and see if it works in theory? Or did you try rescoring training data and train a net on that?
